apps/theming-designer/src/components/FabricSlotWidget.tsx

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import { IThemeRules, FabricSlots, IThemeSlotRule } from 'office-ui-fabric-react/lib/ThemeGenerator';
+import { IColor } from 'office-ui-fabric-react/lib/Color';
+import { Stack } from 'office-ui-fabric-react/lib/Stack';
+import { mergeStyles } from '@uifabric/merge-styles';
+import { ColorPicker } from 'office-ui-fabric-react/lib/ColorPicker';
+import { Callout, DirectionalHint } from 'office-ui-fabric-react/lib/Callout';
+import { Text } from 'office-ui-fabric-react';
+
+export interface IFabricSlotWidgetProps {
+  slot: FabricSlots;
+  onFabricPaletteColorChange: (newColor: IColor | undefined, fabricSlot: FabricSlots) => void;
+  themeRules?: IThemeRules;
+  directionalHint?: DirectionalHint;
+}
+
+export interface IFabricSlotWidgetState {
+  isColorPickerVisible: boolean;
+  colorPickerElement: HTMLElement | null;
+}
+
+const slotClassName = mergeStyles({
+  display: 'flex',
+  alignItems: 'center',
+  overflow: 'auto'
+});
+
+const fabricPaletteColorBox = mergeStyles({
+  width: 15,
+  height: 15,
+  display: 'inline-block',
+  left: 5,
+  top: 5,
+  border: '1px solid black',
+  flexShrink: 0
+});
+
+const colorBoxAndHexStringClassName = mergeStyles({
+  position: 'relative' /* This is necessary to make position: absolute; work in the other style. */
+});
+
+export class FabricSlotWidget extends React.Component<IFabricSlotWidgetProps, IFabricSlotWidgetState> {
+  constructor(props: any) {
+    super(props);
+
+    this.state = {
+      isColorPickerVisible: false,
+      colorPickerElement: null
+    };
+
+    this._onColorBoxClick = this._onColorBoxClick.bind(this);
+    this._onCalloutDismiss = this._onCalloutDismiss.bind(this);
+  }
+
+  public render() {
+    const { isColorPickerVisible, colorPickerElement } = this.state;
+    const slotRule = this.props.themeRules![FabricSlots[this.props.slot]];
+    return (
+      <div key={slotRule.name} className={slotClassName}>
+        <Stack horizontal className={colorBoxAndHexStringClassName} gap={5}>
+          <div
+            key={slotRule.name}
```

Comment:
I'm not sure that you need `key` here. Generally, you only need them for repeated sibling content so React knows what to diff when updates occur.

Reply:
There's a FabricSlotWidget being created for each of the Fabric Slots so it's technically repeated content no?

Reply:
My understanding is this applies if the components are sibling components. Is that the case here? If so then a `key` is appropriate.

>Keys help React identify which items have changed, are added, or are removed. Keys should be given to the elements inside the array to give the elements a stable identity:
>
>The best way to pick a key is to use a string that uniquely identifies a list item among its siblings. Most often you would use IDs from your data as keys:
>
>https://reactjs.org/docs/lists-and-keys.html#keys

Reply:
Hmmm I guess they are all quite individual and not necessarily sibling components.